data/transactions/verify/verifiedTxnCache.go

```diff
@@ -0,0 +1,292 @@
+
+package verify
+
+import (
+	"errors"
+
+	"github.com/algorand/go-deadlock"
+
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+const entriesPerBucket = 8179 // the default bucket size; a prime number could promote a lower hash collisions in case the hash function isn't perfect.
+const maxPinnedEntries = 500000
+
+// VerifiedTxnCacheError helps to identifiy the errors of a cache error and diffrenciate these from a general verification errors.
+type VerifiedTxnCacheError struct {
+	inner error
+}
+
+// Unwrap provides accesss to the underlying error
+func (e *VerifiedTxnCacheError) Unwrap() error {
+	return e.inner
+}
+
+// Error formats the underlying error message
+func (e *VerifiedTxnCacheError) Error() string {
+	return e.inner.Error()
+}
+
+// errTooManyPinnedEntries is being generated when we attempt to pin an transaction while we've already exceeded the maximum number of allows
+// transactions in the verification cache.
+var errTooManyPinnedEntries = &VerifiedTxnCacheError{errors.New("Too many pinned entries")}
+
+// errMissingPinnedEntry is being generated when we're trying to pin a transaction that does not appear in the cache
+var errMissingPinnedEntry = &VerifiedTxnCacheError{errors.New("Missing pinned entry")}
+
+// VerifiedTransactionCache provides a cached store of recently verified transactions
+type VerifiedTransactionCache interface {
+	// Add adds a given transaction group and it's associated group context to the cache. If any of the transactions already appear
+	// in the cache, the new entry overrides the old one.
+	Add(txgroup []transactions.SignedTxn, groupCtx *GroupContext)
+	// AddPayset works in a similar way to Add, but is intended for adding an array of transaction groups, along with their corresponding contexts.
+	AddPayset(txgroup [][]transactions.SignedTxn, groupCtxs []*GroupContext) error
+	// GetUnverifiedTranscationGroups compares the provided payset against the currently cached transactions and figure which transaction groups aren't fully cached.
+	GetUnverifiedTranscationGroups(payset [][]transactions.SignedTxn, CurrSpecAddrs transactions.SpecialAddresses, CurrProto protocol.ConsensusVersion) [][]transactions.SignedTxn
+	// UpdatePinned replaces the pinned entries with the one provided in the pinnedTxns map. This is typically expected to be a subset of the
+	// already-pinned transactions. If a transaction is not currently pinned, and it's can't be found in the cache, a errMissingPinnedEntry error would be generated.
+	UpdatePinned(pinnedTxns map[transactions.Txid]transactions.SignedTxn) error
+	// Pin pins the given transaction group.
+	Pin(txgroup []transactions.SignedTxn) error
+}
+
+// VerifiedTransactionCacheImpl provides an implementation of the VerifiedTransactionCache interface
+type verifiedTransactionCacheImpl struct {
```

Comment:
I used to Impl suffixes after C++ but in Go `verifiedTransactionCache` looks as a proper naming